### PR TITLE
Scala.js: Implement JS exports.

### PR DESCRIPTION
**Pull request by sjrd**
Compared to the implementation in Scala 2, there is one significant difference. In Scala 2, top-level exports and static exports info are computed by `PrepJSInterop` and stored in an out-of-band mutable Map until the back-end. In this implementation, `PrepJSInterop` only performs checks for compile error messages, and the required data are recomputed by the back-end from the annotations.

Since dotc does not (currently?) constant-fold expressions within annotations, `PrepJSInterop` specifically constant-folds the arguments of `@JSExportTopLevel` and `@JSExportStatic`, so that the literal values can be recovered by the back-end.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala [hunk 1]
```diff
@@ -147,6 +147,14 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer { thisPhase
...
           checkInferredWellFormed(tree.tpt)
           val sym = tree.symbol
+          if sym.is(Method) then
+            if sym.isSetter then
+              removeUnwantedAnnotations(sym, defn.SetterMetaAnnot, NoSymbol, keepIfNoRelevantAnnot = false)
+          else
+            if sym.is(Param) then
+              removeUnwantedAnnotations(sym, defn.ParamMetaAnnot, NoSymbol, keepIfNoRelevantAnnot = true)
+            else
+              removeUnwantedAnnotations(sym, defn.GetterMetaAnnot, defn.FieldMetaAnnot, keepIfNoRelevantAnnot = !sym.is(ParamAccessor))
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/PostTyper.scala:157]: Thanks for implementing this stuff! This should fix https://github.com/lampepfl/dotty/issues/9881

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala:157]: Ah well, I didn't really have a choice. I had tests that needed to pass. 🙃

**Comment by sjrd** [comment]
I added a commit that adds neg tests and makes them pass. It should be squashed eventually, so if you haven't started reviewing, it's better to review the PR diff than the commits separately.

File: compiler/src/dotty/tools/dotc/transform/PostTyper.scala (hunk 1 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala:157]: Can you add the testcase from that issue to make sure it's actually fixed?

Comment (sjrd) [reply compiler/src/dotty/tools/dotc/transform/PostTyper.scala:157]: Done.

File: compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala [hunk 2]
```diff
@@ -42,6 +43,254 @@ final class JSExportsGen(jsCodeGen: JSCodeGen)(using Context) {
   import jsCodeGen._
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala:43]: I noticed the scala 2 implementation is named `GenJSExports` and this one is named `JSExportsGen`. Was this change intentional?

**Review by nicolasstucki — APPROVED** [review]

File: compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala (hunk 2 above)
Comment (sjrd) [reply compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala:43]: Yes, it was more consistent with `JSCodeGen`.

**Comment by sjrd** [comment]
Rebased, squashed, and add `prepjsinterop` in the phases to `Ycheck`, now that #10116 is fixed.

**Review by nicolasstucki — APPROVED** [review]

File: compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala [hunk 3]
```diff
@@ -3916,19 +3933,17 @@ class JSCodeGen()(using genCtx: Context) {
       }
 
       (f, true)
-    } else /*if (jsInterop.topLevelExportsOf(sym).nonEmpty) {
-      val f = js.SelectStatic(encodeClassName(sym.owner),
-          encodeFieldSym(sym))(jstpe.AnyType)
+    } else if (sym.hasAnnotation(jsdefn.JSExportTopLevelAnnot)) {
```
Comment (gzm0) [inline compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala:3936]: Do you recall why we didn't simply do this in Scala 2?

**Review by gzm0 — APPROVED** [review]

File: compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala (hunk 3 above)
Comment (sjrd) [reply compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala:3936]: I don't know. It's probably historical, following a long chain of changes. `@JSExportTopLevel` inherited from `@JSExport` on classes and objects, in which the name could be implicit; which might explain it. git blame suggests https://github.com/scala-js/scala-js/commit/273e9b82fea10628e4669aba121027b07c34f5e0 as the moment at which we introduced those out-of-the-band mutable maps, but at the time `@JSExportTopLevel` did not exist yet.

We could try to get rid of it in Scala 2.
